core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketsort/BucketSortPipelineAggregator.java

```diff
@@ -0,0 +1,192 @@
+        out.writeVInt(from);
+        out.writeOptionalVInt(size);
+        gapPolicy.writeTo(out);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return BucketSortPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext reduceContext) {
+        InternalMultiBucketAggregation<InternalMultiBucketAggregation, InternalMultiBucketAggregation.InternalBucket> originalAgg =
+                (InternalMultiBucketAggregation<InternalMultiBucketAggregation, InternalMultiBucketAggregation.InternalBucket>) aggregation;
+        List<? extends InternalMultiBucketAggregation.InternalBucket> buckets = originalAgg.getBuckets();
+        int bucketsCount = buckets.size();
+        int offset = reduceContext.isFinalReduce() ? from : 0;
+        int currentSize = size == null ? bucketsCount : size;
+
+        if (offset >= bucketsCount) {
+            return originalAgg.create(Collections.emptyList());
+        }
+
+        int resultSize = Math.min(currentSize, bucketsCount - offset);
+
+        // If no sorting needs to take place, we just truncate and return
+        if (sorts.size() == 0) {
+            return originalAgg.create(truncate(buckets, offset, currentSize));
+        }
+
+        int queueSize = Math.min(offset + currentSize, bucketsCount);
+        PriorityQueue<ComparableBucket> ordered = new TopNPriorityQueue(queueSize);
+        for (InternalMultiBucketAggregation.InternalBucket bucket : buckets) {
+            ordered.insertWithOverflow(new ComparableBucket(originalAgg, bucket));
+        }
+
+        // Popping from the priority queue returns the least element. The elements we want to skip due to offset would pop last.
+        // Thus, we just have to pop as many elements as we expect in results and store them in reverse order.
+        LinkedList<InternalMultiBucketAggregation.InternalBucket> newBuckets = new LinkedList<>();
+        for (int i = resultSize - 1; i >= 0; --i) {
+            ComparableBucket comparableBucket = ordered.pop();
+            if (comparableBucket.skip == false) {
+                newBuckets.addFirst(comparableBucket.getInternalBucket());
+            }
+        }
+        return originalAgg.create(newBuckets);
+    }
+
+    private static List<InternalMultiBucketAggregation.InternalBucket> truncate(
+            List<? extends InternalMultiBucketAggregation.InternalBucket> buckets, int offset, int size) {
+
+        List<InternalMultiBucketAggregation.InternalBucket> truncated = new ArrayList<>(size);
+        for (int i = offset; i < offset + size; ++i) {
+            truncated.add(buckets.get(i));
+        }
+        return truncated;
+    }
+
+    private class ComparableBucket implements Comparable<ComparableBucket> {
+
+        private final MultiBucketsAggregation parentAgg;
+        private final InternalMultiBucketAggregation.InternalBucket internalBucket;
+        private boolean skip = false;
```

Comment:
Could you add a JavaDoc to explain what this is for?

Reply:
Done